flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -98,38 +93,66 @@ public ResourceAllocationResult tryFulfillRequirements(
-                                TaskManagerInfo::getInstanceId,
-                                taskManager ->
-                                        Tuple2.of(
-                                                taskManager.getAvailableResource(),
-                                                taskManager.getDefaultSlotResourceProfile())));
+                .map(
+                        taskManager ->
+                                new InternalResourceInfo(
+                                        taskManager.getInstanceId(),
+                                        taskManager.getDefaultSlotResourceProfile(),
+                                        taskManager.getAvailableResource(),
+                                        resultBuilder))
+                .collect(Collectors.toList());
     }
 
-    private static Map<PendingTaskManagerId, ResourceProfile> getPendingResources(
-            TaskManagerResourceInfoProvider taskManagerResourceInfoProvider) {
+    private static List<InternalResourceInfo> getPendingResources(
+            TaskManagerResourceInfoProvider taskManagerResourceInfoProvider,
+            ResourceAllocationResult.Builder resultBuilder) {
         return taskManagerResourceInfoProvider.getPendingTaskManagers().stream()
-                .collect(
-                        Collectors.toMap(
-                                PendingTaskManager::getPendingTaskManagerId,
-                                PendingTaskManager::getTotalResourceProfile));
+                .map(
+                        pendingTaskManager ->
+                                new InternalResourceInfo(
+                                        pendingTaskManager.getPendingTaskManagerId(),
+                                        pendingTaskManager.getDefaultSlotResourceProfile(),
+                                        pendingTaskManager.getTotalResourceProfile(),
+                                        resultBuilder))
+                .collect(Collectors.toList());
+    }
+
+    private static int tryFulfilledRequirementWithResource(
+            List<InternalResourceInfo> internalResource,
+            int numUnfulfilled,
+            ResourceProfile requiredResource,
+            JobID jobId) {
+        final Iterator<InternalResourceInfo> internalResourceInfoItr = internalResource.iterator();
+        while (numUnfulfilled > 0 && internalResourceInfoItr.hasNext()) {
+            final InternalResourceInfo currentTaskManager = internalResourceInfoItr.next();
+            while (numUnfulfilled > 0
+                    && currentTaskManager.tryAllocateSlotForJob(jobId, requiredResource)) {
+                numUnfulfilled--;
+            }
+            if (currentTaskManager.availableProfile.equals(ResourceProfile.ZERO)) {
+                internalResourceInfoItr.remove();
+            }
+        }
+        return numUnfulfilled;
     }
 
     private static ResourceCounter tryFulfillRequirementsForJobWithRegisteredResources(
```

Comment:
This can also be deduplicated, if we break matching existing pending resources and allocating new pending resources into two loops.